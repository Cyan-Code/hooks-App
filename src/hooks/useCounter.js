import { useState } from "react"

export const useCounter = (inicialState = 10 ) => {

  const [state, setState] = useState(inicialState)

  const increment = (factor = 2) => {
    setState( state + factor )
  }

  const decrement = (factor = 2) => {
    setState( state - factor)
  }

  const reset = () => {
    setState( inicialState )
  }
  
  return {
    state,
    increment,
    decrement,
    reset
  }

}
 