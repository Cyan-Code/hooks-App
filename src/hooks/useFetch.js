import { useEffect, useRef, useState } from "react"

export const useFetch = ( url ) => {
  const [state, setState] = useState({data: null, loading: true, error: null})

  const isMounted = useRef(true) //

  useEffect( () => { // Cuando este efecto se desmonte, cambia el estado de referencia
    return () => { // no re-renderiza el componente, simplemente mantiene una referencia
      isMounted.current = false // basicamente, es falso cuando deja de existir
    }
  }, [])

  useEffect( () => {

    setState({data: null, loading: true, error: null})

    fetch( url )
      .then( resp => resp.json() )
      .then( data => {
        if (isMounted.current) {
          setState ({ // gracias a las validaciones antes hechas con el use ref, se puede hacer
            loading: false, // un trabajo condicional con este componente, ya que se puede
            error: null, // trabajar de manera segura si esta montado, y no necesariamente montarlo
            data // y llamar el useState, si no se encuentra montado
          })
        }  
      })
  }, [url])

  return state
}
