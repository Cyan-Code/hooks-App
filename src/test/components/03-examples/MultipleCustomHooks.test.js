import React from "react";
import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";

jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

describe("Pruebas sobre MultipleCustomHooks", () => {
  test("Debe de mostrarse correctamente", () => {
    useCounter.mockReturnValue({ 
      counter: 10,
      increment: () => {}
    })    
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    })
    const wrapper = shallow(<MultipleCustomHooks />);
    expect( wrapper ).toMatchSnapshot()
  });

  test('Debe de mostrar la informacion', () => {
    useCounter.mockReturnValue({ 
      counter: 10,
      increment: () => {}
    })
    useFetch.mockReturnValue({
      data: [{
        author: 'Luis',
        quote: 'Hola Mundo'
      }],
      loading: false,
      error: null
    })
    const wrapper = shallow(<MultipleCustomHooks />)
    expect( wrapper.find('.alert').exists() ).toBe(false)
    expect( wrapper.find('.mb-0').text().trim() ).toBe('Hola Mundo')
    expect( wrapper.find('footer').text().trim() ).toBe('Luis')
  })
});
