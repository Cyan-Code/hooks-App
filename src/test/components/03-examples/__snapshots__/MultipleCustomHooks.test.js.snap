// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre MultipleCustomHooks Debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    Breaking Bad with Multiple hooks
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
     siguiente 
  </button>
</div>
`;
